### PR TITLE
Zenith experiments

### PR DESCRIPTION
**Pull request by lubennikovaav**
PR to have a place for comments and discussion.

### CODE REVIEW

File: README.zenith [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+4. Act as a page cache, so that the primary can request pages from the
+storage nodes, instead of restoring whole relation files from S3.
+
+TODO in Postgres code
+------------------------
+
+- Currently, a relation can be in one of two states: a normal relation, kept in
+local disk, and checkpointed and updated in the primary as normal. Or lazy,
+meaning it hasn't been restored from S3 yet. Introduce third state, where pages
+are fetched from storage nodes as needed, and not written to disk locally.
```
Comment (lubennikovaav) [inline README.zenith:75]: Do really need a third state of relation? Why keeping lazy relations "always lazy" is not enough? We will never checkpoint such relations and will refresh them by pulling data from the storage node on-demand.

File: README.zenith [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+- Local cache invalidation of read-only nodes, so that a read-only node can
+follow a primary, possibly with some lag. Currently, read-only nodes are
+frozen at a particular LSN when they're created.
```
Comment (lubennikovaav) [inline README.zenith:79]: Do you mean buffer cache here?
A very naive approach to advancing LSN is to check each page as we read it from the buffer cache. And if its LSN is less than the newly set LSN for the node, we should request this page (or whole segment ?) from the storage node. Of course, we can cache this info about the last cached LSN for each relation, i.e.in SMgrRelationData.

File: README.zenith (hunk 1 above)
Comment (hlinnaka) [reply README.zenith:75]: The third state is for operating with the Page Servers. With a Page Server, you don't want to fetch the whole 1 GB segment, just the page you need. @kelvich does that vibe with your ideas for the Page Server?

File: README.zenith (hunk 2 above)
Comment (hlinnaka) [reply README.zenith:79]: > Do you mean buffer cache here?

Yeah.

> A very naive approach to advancing LSN is to check each page as we read it from the buffer cache. And if its LSN is less than the newly set LSN for the node, we should request this page (or whole segment ?) from the storage node. Of course, we can cache this info about the last cached LSN for each relation, i.e.in SMgrRelationData.

Hmm. I think that's the same as completely resetting the buffer cache every time the current LSN is advanced? Sure, that will do for the first version.

File: README.zenith (hunk 1 above)
Comment (kelvich) [reply README.zenith:75]: Yes, I was thinking of two kind of queries from processing node smgr to the page store:
1) get one page
2) get stream of pages for some relation (for seq scans)

**Pull request closed without merging**